presto-main/src/main/java/io/prestosql/operator/scalar/FormatFunction.java

```diff
@@ -0,0 +1,302 @@
+        }
+        else if (type.getJavaType() == double.class) {
+            function = (session, block) -> type.getDouble(block, position);
+        }
+        else if (type.getJavaType() == boolean.class) {
+            function = (session, block) -> type.getBoolean(block, position);
+        }
+        else if (type.getJavaType() == Slice.class) {
+            function = (session, block) -> type.getSlice(block, position);
+        }
+        else {
+            function = (session, block) -> type.getObject(block, position);
+        }
+
+        MethodHandle handle = castToVarchar(functionRegistry, type);
+        if ((handle == null) || (handle.type().parameterCount() != 1)) {
+            throw new PrestoException(NOT_SUPPORTED, "Type not supported for formatting: " + type.getDisplayName());
+        }
+
+        return (session, block) -> convertToString(handle, function.apply(session, block));
+    }
+
+    private static MethodHandle castToVarchar(FunctionRegistry functionRegistry, Type type)
+    {
+        try {
+            Signature cast = functionRegistry.getCoercion(type, VARCHAR);
+            return functionRegistry.getScalarFunctionImplementation(cast).getMethodHandle();
+        }
+        catch (OperatorNotFoundException e) {
+            return null;
+        }
+    }
+
+    private static ZonedDateTime toZonedDateTime(long value)
+    {
+        Instant instant = Instant.ofEpochMilli(unpackMillisUtc(value));
+        ZoneId zoneId = ZoneId.of(unpackZoneKey(value).getId());
+        return ZonedDateTime.ofInstant(instant, zoneId);
+    }
+
+    private static LocalDateTime toLocalDateTime(ConnectorSession session, long value)
+    {
+        Instant instant = Instant.ofEpochMilli(value);
+        if (session.isLegacyTimestamp()) {
+            ZoneId zoneId = ZoneId.of(session.getTimeZoneKey().getId());
+            return LocalDateTime.ofInstant(instant, zoneId);
+        }
+        return LocalDateTime.ofInstant(instant, ZoneOffset.UTC);
+    }
+
+    private static LocalTime toLocalTime(ConnectorSession session, long value)
+    {
+        if (session.isLegacyTimestamp()) {
+            Instant instant = Instant.ofEpochMilli(value);
+            ZoneId zoneId = ZoneId.of(session.getTimeZoneKey().getId());
+            return ZonedDateTime.ofInstant(instant, zoneId).toLocalTime();
+        }
+        return LocalTime.ofNanoOfDay(MILLISECONDS.toNanos(value));
+    }
+
+    private static Object convertToString(MethodHandle handle, Object value)
+    {
+        try {
+            return ((Slice) handle.invoke(value)).toStringUtf8();
+        }
+        catch (Throwable t) {
+            throw internalError(t);
+        }
+    }
+
+    public static final class NoRow
+    {
+        private NoRow() {}
+
+        @ScalarFunction(value = NAME_NO_ROW, hidden = true)
+        @SqlType(StandardTypes.VARCHAR)
+        public static Slice noArgs(ConnectorSession session, @SqlType(StandardTypes.VARCHAR) Slice format)
```

Comment:
I don't think we should support a version with no args. A format string with no placeholders is effectively a string constant, so `format('xxxx') = 'xxxx'`

Reply:
I was thinking it could make certain code generation scenarios simpler, but I agree it's better not to support this.